herddb-core/src/main/java/herddb/core/TableManager.java

```diff
@@ -1127,28 +1147,54 @@ private LockHandle lockForRead(Bytes key, Transaction transaction) {
                     .exception(new RecordTooBigException("New record " + key + " is to big to be inserted: size " + size + ", max size " + maxLogicalPageSize));
         }
 
-        LockHandle lock = lockForWrite(key, transaction);
         CompletableFuture<StatementExecutionResult> res = null;
+        LockHandle lock = null;
+        try {
+            lock = lockForWrite(key, transaction);
+            if (uniqueIndexes != null) {
+                for (UniqueIndexLockReference uniqueIndexLock : uniqueIndexes) {
+                    AbstractIndexManager index = uniqueIndexLock.indexManager;
+                    LockHandle lockForIndex = lockForWrite(uniqueIndexLock.key, transaction, index.getIndexName(), index.getLockManager());
+                    if (transaction == null) {
+                        uniqueIndexLock.lockHandle = lockForIndex;
+                    }
+                    if (index.valueAlreadyMapped(uniqueIndexLock.key, null)) {
```

Comment:
This procedure will require a double index search. Can we posticipate to the actual insertion?

Reply:
No we can't.
It is the same flow as for the PK.

Actual writes will happen after writing to the log. We can't do best